src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -367,10 +378,54 @@ public byte[] Http2AltSvcOriginUri
             }
         }
 
-        public bool EnableMultipleHttp2Connections => _poolManager.Settings.EnableMultipleHttp2Connections;
+        private bool EnableMultipleHttp2Connections => _poolManager.Settings.EnableMultipleHttp2Connections;
 
         /// <summary>Object used to synchronize access to state in the pool.</summary>
-        private object SyncObj => _idleConnections;
+        private object SyncObj
+        {
+            get
+            {
+                Debug.Assert(!Monitor.IsEntered(_availableHttp11Connections));
```

Comment:
What is wrong with reentering the same lock? Is it a performance hit? We don't do anything similar anywhere else as far as I can see, so I'm curious.

Reply:
It can lead to issues where you assume you've released the lock, but you haven't actually done so, which can lead to deadlock.

It can also lead to data consistency issues. For example, you hold the lock and so you assume the data that is protected by the lock can't change. Then you call some other method, which takes the lock reentrantly and modifies the data -- after all it thinks this is safe, because it has the lock. But now the data has changed underneath the original caller, which is usually not expected and can lead to weird issues.

I'm not always hardcore about protecting against reentrancy like this, but when it becomes hard to reason about all the places where the lock can be taken, I usually add something like this to protect against reentrancy.